### ISSUE TITLE
Cannot parse RFC3339 with second precision

### ISSUE THREAD

**Issue by harryrose** [sitewhere/sitewhere#85]
Hi,

Version: 1.8.0 (via the docker hub)

There appears to be an issue while parsing RFC3339 dates in series requests when they are missing the millisecond component.  While making the following GET request I get a 500 response from the sitewhere service:

```
GET http://localhost:8080/sitewhere/api/assignments/b9470380-62bc-49e9-a08c-0192d2bbc532/measurements/series?endDate=2016-12-23T01%3A00%3A00Z&startDate=2016-12-23T00%3A00%3A00Z
````

Response:
```
{
  "timestamp": 1482508720964,
  "status": 500,
  "error": "Internal Server Error",
  "exception": "org.springframework.web.method.annotation.MethodArgumentTypeMismatchException",
  "message": "Failed to convert value of type [java.lang.String] to required type [java.util.Date]; nested exception is org.springframework.core.convert.ConversionFailedException: Failed to convert from type [java.lang.String] to type [@com.wordnik.swagger.annotations.ApiParam @org.springframework.web.bind.annotation.RequestParam @org.springframework.format.annotation.DateTimeFormat java.util.Date] for value '2016-12-23T00:00:00Z'; nested exception is java.lang.IllegalArgumentException: Invalid format: \"2016-12-23T00:00:00Z\" is malformed at \"Z\"",
  "path": "/sitewhere/api/assignments/b9470380-62bc-49e9-a08c-0192d2bbc532/measurements/series"
}
```

When making the request with milliseconds specified, the query succeeds with a 200 status.

i.e.

```
http://localhost:8080/sitewhere/api/assignments/b9470380-62bc-49e9-a08c-0192d2bbc532/measurements/series?endDate=2016-12-23T01%3A00%3A00.000Z&startDate=2016-12-23T00%3A00%3A00.000Z
```

I suspect the inability to parse the date without millis is a bug, but if not, I think a 400 response would be more usual than a 500.

**Comment by derekadams** [comment]
Thanks for reporting the issue. The issue shows up in all of the REST services that take dates as parameters. We added a Jira issue to get this fixed for the 1.10.0 release.

https://sitewhere.atlassian.net/browse/SITEWHERE-369

**Comment by jmrubillon** [comment]
It would appear that higher precision than ms timestamps also cause an issue although more subtle.
A message like the following JSON message:
{"hardwareId": "nim-wsn-13", "measurements": [{"measurements": {"temperature": 25,"light": 727,"humidity": 98},"eventDate": "2017-02-14T10:34:30.6160615Z"}]}
gets stored in InfluxDB as a timestamp of 1487074630615000000 which corresponds to: '2017-02-14T12:17:11Z' which clearly isn't the same time.
Truncating the eventDate to "2017-02-14T10:34:30.616Z" works fine but loses the nanosecond precision offered by the underlying datastore.

**Comment by garysmi2** [comment]
Just to point out the precision problem as stated by jmrubillon using InfluxDBEvents, part of the problem looks like Joda time  which can only handle up to the millisecond (three digits), Joda DateTimeFormatterBuilder has a method:

```
appendFractionOfSecond(int minDigits, int maxDigits)
Appends the print/parse of a fractional second.
```
which may be of use and looks like java 8 time supports up to 9 digits

**Comment by garysmi2** [comment]
And looking at it again, looks like the assumed precision is milliseconds everywhere. 

I can work around this in influxdb at least by setting the time precision as NanoSeconds when building the event and sending the eventDate as a timestamp rather than a human readable string but this causes precision errors when in the reverse case. Possibly making the precision configurable via event meta-data for influxdb might be a solution, anyway, I've hi-jacked this thread enough :)

**Comment by derekadams** [comment]
I think we use java Date objects for the most part in dealing with time values for events, but there may be cases where we do conversions to "milliseconds since epoch" in doing calculations. I know for sure that the HBase event storage code only supports a granularity of milliseconds for now. We will have to take a look at the end-to-end handling of the event and see if anything prevents using nanosecond precision.

Regarding choosing the InfluxDB time precision, that may be a good candidate for a setting on the InfluxDB configuration page since it would most likely be a setting that applies to all event data.

**Comment by jmrubillon** [comment]
For InfluxDB I would not have a setting common to every measurement that comes in.
As per InfluxDB's documentation, the granularity of measurements should be as coarse as possible to optimise performance. It is better to have a part of the event message (optional field) that states what the precision of the eventDate field is.
I believe that's what @garysmi2 code change provides. If the precision is not stated, then the current default of millisecond precision is assumed.

**Comment by derekadams** [comment]
Ok guys, sorry for the late response. I checked the fix into the 1.10.0 branch. It should handle pretty much any ISO-style date now.
